test/functional/options/defaults_spec.lua

```diff
@@ -328,3 +332,235 @@ describe('XDG-based defaults', function()
+    end)
+
+    it("doesn't expand $VARIABLES", function()
+      clear({env={
+        XDG_DATA_HOME='$VARIABLES',
+        VARIABLES='this-should-not-happen'
+      }})
+      eq(alter_path('$VARIABLES' .. appended_dir), funcs.stdpath('data'))
+    end)
+
+    it("doesn't expand ~/", function()
+      clear({env={
+        XDG_DATA_HOME='~/frobnitz'
+      }})
+      eq(alter_path('~/frobnitz' .. appended_dir), funcs.stdpath('data'))
+    end)
+  end)
+
+  describe('"cache"' , function ()
+    it('knows XDG_CACHE_HOME', function()
+      clear({env={
+        XDG_CACHE_HOME=alter_path('/home/docwhat/.cache')
+      }})
+      eq(alter_path('/home/docwhat/.cache/nvim'), funcs.stdpath('cache'))
+    end)
+
+    it('handles changes during runtime', function()
+      clear({env={
+        XDG_CACHE_HOME=alter_path('/home/original')
+      }})
+      eq(alter_path('/home/original/nvim'), funcs.stdpath('cache'))
+      nvim('command', 'let $XDG_CACHE_HOME="/home/new"')
+      eq(alter_path('/home/new/nvim'), funcs.stdpath('cache'))
+    end)
+
+    it("doesn't expand $VARIABLES", function()
+      clear({env={
+        XDG_CACHE_HOME='$VARIABLES',
+        VARIABLES='this-should-not-happen'
+      }})
+      eq(alter_path('$VARIABLES/nvim'), funcs.stdpath('cache'))
+    end)
+
+    it("doesn't expand ~/", function()
+      clear({env={
+        XDG_CACHE_HOME='~/frobnitz'
+      }})
+      eq(alter_path('~/frobnitz/nvim'), funcs.stdpath('cache'))
+    end)
+  end)
+
+  describe('"config_dirs"' , function ()
+    it('knows XDG_CONFIG_DIRS', function()
+      clear({env={
+        XDG_CONFIG_DIRS=alter_path('/home/docwhat/.config:/etc/config')
+      }})
+      eq({
+        alter_path('/home/docwhat/.config/nvim'),
+        alter_path('/etc/config/nvim'),
+      }, funcs.stdpath('config_dirs'))
+    end)
+
+    it('handles changes during runtime', function()
+      clear({env={
+        XDG_CONFIG_DIRS=alter_path('/home/original:/etc/config')
+      }})
+      eq({
+        alter_path('/home/original/nvim'),
+        alter_path('/etc/config/nvim')
+      }, funcs.stdpath('config_dirs'))
+      nvim('command', 'let $XDG_CONFIG_DIRS="/home/new:/etc/config"')
+      eq({
+        alter_path('/home/new/nvim'),
+        alter_path('/etc/config/nvim')
+      }, funcs.stdpath('config_dirs'))
+    end)
+
+    it("doesn't expand $VAR and $IBLES", function()
+      clear({env={
+        XDG_CONFIG_DIRS='$VAR:$IBLES',
+        VAR='this-should-not-happen',
+        IBLES='neither-should-this'
+      }})
+      eq({
+        alter_path('$VAR/nvim'),
+        alter_path('$IBLES/nvim'),
+      }, funcs.stdpath('config_dirs'))
+    end)
+
+    it("doesn't expand ~/", function()
+      clear({env={
+        XDG_CONFIG_DIRS='~/.oldconfig:~/.olderconfig'
+      }})
+      eq({
+        alter_path('~/.oldconfig/nvim'),
+        alter_path('~/.olderconfig/nvim'),
+      }, funcs.stdpath('config_dirs'))
+    end)
+  end)
+
+  describe('"data_dirs"' , function ()
```

Comment:
Missing returns in cases “colon-separated variable contains one directory”, “colon-separated variable contains no directories” (do not forget to set $HOME, $LOCALAPPDATA and $TEMP), “colon-separated variable contains item larger then MAXPATHL” (e.g. 8192 bytes). And I would highly suggest to create a test generator function like

```lua
local home = …  -- depends on OS
local localappdata = ...  -- Something on Windows, empty string on *nix
local function test_colon_env(msg, stdpath_arg, env_name, env_val, expected_paths, env_val_runtime, expected_paths_runtime)
  it(msg, function()
    clear({env={[env_name]=env_val, HOME=home, LOCALAPPDATA=localappdata}})
    eq(expected_paths, funcs.stdpath(stdpath_arg))

    if not env_val_runtime then return end
    meths.set_var('env_val', env_val_runtime)
    command(('let $%s=g:env_val'):format(env_name))
    eq(expected_paths_runtime, funcs.stdpath(stdpath_arg))
  end)
end
```

Using such generators makes code less verbose.